course02/homework02/Tomasz/message_throttler.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include<memory>
+
+template<
+	typename _ClientId,
+	typename _Message,
+	typename _MessageConsumer,
+	typename _MessageDisposer,
+	typename _Timestamp,
+	typename _Timestamper,
+	typename _TimestampThreshold,
+	typename _MessageThrottlerInterface,
+	typename _Map
+>
+class message_throttler
+{
+public:
+	message_throttler(
+		const _MessageConsumer& messageConsumer, 
+		const _MessageDisposer& messageDisposer, 
+		const _Timestamper& timestamper,
+		const _TimestampThreshold& timestampThreshold) :
+		mMessageConsumer(messageConsumer),
+		mMessageDisposer(messageDisposer),
+		mTimestamper(timestamper),
+		mTimestampThreshold(timestampThreshold)
+	{ }
+
+	_MessageThrottlerInterface& from(const _ClientId& clientId)
+	{
+		auto element = mClients.emplace(clientId, nullptr);
+		auto& client = element.first->second;
+
+		if (element.second)
+			client = std::make_unique<_MessageThrottlerInterface>(mMessageConsumer, mMessageDisposer, mTimestamper, mTimestampThreshold);
```

Comment:
I find the pattern of inserting a nullptr, and modifying it to a unique_ptr if inserted is a bit complicated. Why not just

```
auto& client = mClients[clientId];
if (client == nullptr)
    client = std::make_unique...
return client;
```

Although I must admit that I prefer an APIwhere clients would be explicitly announced through another function, and her we would always expect to `find` a client without creating it at the same time

Reply:
Yeah, the history of this line is quite old :P 
You are completely, this pattern is much cleaner.